File: 1170-compare-strings-by-frequency-of-the-smallest-character/1170-compare-strings-by-frequency-of-the-smallest-character.py
class Solution:
    def __init__(self):
        self.d = {}
    def numSmallerByFrequency(self, queries: List[str], words: List[str]) -> List[int]:
        words = sorted(words,key = lambda word: self.getValue(word))
        output = [0] * len(queries)
        for i,query in enumerate(queries):
            output[i] = len(words) - self.find(words,self.getValue(query))
        return output
    
    def find(self,words,limit):
        low = 0
        high = len(words) - 1
        
        while low < high:
            mid = low + (high - low) // 2
            
            if self.getValue(words[mid]) > limit:
                high = mid
            else:
                low = mid + 1
        if low == high and self.getValue(words[low]) <= limit:
            return len(words)
        return low
    def getValue(self,s):
        if s in self.d:return self.d[s]
        lowest = 123
        freq = 0
        for char in s:
            if ord(char) < lowest:
                lowest = ord(char)
                freq = 1
            if ord(char) == lowest:
                freq += 1
        self.d[s] = freq
        return freq
        
        
    
        

File: 1170-compare-strings-by-frequency-of-the-smallest-character/1170-compare-strings-by-frequency-of-the-smallest-character.java
class Solution {
    public int[] numSmallerByFrequency(String[] queries, String[] words) {
        Arrays.sort(words,(w1,w2)->getValue(w1)-getValue(w2));

        int [] output = new int[queries.length];
            int i = 0;
            for (String query:queries){
                output[i] = words.length - find(words,getValue(query));
                    i++;
            }
        return output;
    }
    int find(String [] words,int bottomLimit){
        
        int low = 0;
        int high = words.length - 1;
        
        while (low < high){
            int mid = low + (high - low ) / 2;
            
            if (getValue(words[mid]) > bottomLimit)
                high = mid;
            else low = mid + 1;
        }
        if (low == high && getValue(words[low]) <= bottomLimit)
            return words.length;
        
        return low;
    }
    HashMap<String,Integer> map = new HashMap<>();
    int getValue(String s){
        if(map.containsKey(s))return map.get(s);
        char lowest = 'z' + 1;
        int freq = 0;
        for(int i = 0;i < s.length();i++){
            if(s.charAt(i) < lowest){
                lowest = s.charAt(i);
                freq = 1;
            }else if (s.charAt(i) == lowest)
                freq ++;
        }
        map.put(s,freq);
        return freq;
    }
}